DistroLauncher/OOBE.cpp

```diff
@@ -0,0 +1,121 @@
+//  the Free Software Foundation, either version 3 of the License.
+//  .
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+//  GNU General Public License for more details.
+//  .
+//  You should have received a copy of the GNU General Public License
+//  along with this program.If not, see < https://www.gnu.org/licenses/>.
+//  .
+//  On Debian systems, the complete text of the GNU General
+//  Public License version 3 can be found in "/usr/share/common-licenses/GPL-3".
+//
+
+// Holds together the user information retrieved from Win32 API's.
+
+#include "stdafx.h"
+
+namespace DistributionInfo {
+
+	namespace { // Because I don't like exporting what seems to be internal.
+		// Saves Windows information to supply to the OOBE.
+		// Returns empty string if we fail to generate the prefill file
+		// or the postfix to be added to the OOBE command line.
+		std::wstring PreparePrefillInfo() {
+			std::wstring commandLine;
+			std::wstring prefillInfo = DistributionInfo::GetPrefillInfoInYaml();
+			if (!prefillInfo.empty()) {
+				// Write it to a file inside \\wsl$ distro filesystem.        
+				const std::wstring prefillFileNameDest = L"/var/tmp/prefill-system-setup.yaml";
+				const std::wstring wslPrefix = L"\\\\wsl$\\" + DistributionInfo::Name;
+				std::wofstream prefillFile;
+				// Glad that mixing slashes and backslashes that way is not a problem in Windows.
+				prefillFile.open(wslPrefix + prefillFileNameDest, std::ios::ate);
+				prefillFile << prefillInfo;
+				prefillFile.close();
+				if (!prefillFile.fail()) // updating the command line if file was written.
+					commandLine += L" --prefill=" + prefillFileNameDest;
+			}
+			return commandLine;
+		} // std::wstring DistributionInfo::PreparePrefillInfo().
+
+		
+	} // namespace.
+
+	ULONG DistributionInfo::OOBE()
+	{
+		ULONG uid = UID_INVALID;
+
+		// calling the oobe experience
+		DWORD exitCode;
+
+		// Prepare prefill information to send to the OOBE.
+		std::wstring prefillCLIPostFix = DistributionInfo::PreparePrefillInfo();
+
+		// Flutter installer is called ubuntu_wsl_setup in its repo. I'm assuming it will be found inside PATH
+		std::wstring commandLine = DistributionInfo::OOBE_NAME + prefillCLIPostFix;
+
+		HRESULT hr = g_wslApi.WslLaunchInteractive(commandLine.c_str(), true, &exitCode);
+		if ((FAILED(hr)) || (exitCode != 0)) {
+			return uid;
+		}
+
+		hr = g_wslApi.WslLaunchInteractive(L"clear", true, &exitCode);
+		if (FAILED(hr)) {
+			return uid;
+		}
+
+		// getting username from ouput
+		// Create a pipe to read the output of the launched process.
+		HANDLE readPipe;
+		HANDLE writePipe;
+		SECURITY_ATTRIBUTES sa{ sizeof(sa), nullptr, true };
+		if (CreatePipe(&readPipe, &writePipe, &sa, 0)) {
+			// Query the UID of the supplied username.
+			std::wstring command = L"cat /var/lib/ubuntu-wsl/assigned_account";
```

Comment:
Is that part still needed? I think subiquity is now filing wsl.conf with the correct user name directly, which should remove the need for this hack. Am I correct @patrick330602?